ttljob: make TTL work for multi-tenant

otan opened the pull request:
Use the appropriate KV abstraction to make ttl work for serverless
clusters.

We can't use `SPLIT AT` with multi-tenant tests, so we are skipping
those for now.

Release note (sql change): The cluster setting
`sql.ttl.range_batch_size` is deprecated.

Resolves #82386

Release note (sql change): Row-level TTL now works for multi-tenant
clusters.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82390)
<!-- Reviewable:end -->

ajstorm commented:
> We can't use SPLIT AT with multi-tenant tests, so we are skipping
those for now.

@rharding6373 FYI.

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82390)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @otan)

___
*[`pkg/sql/ttl/ttljob/main_test.go` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82390#-N3f96VR84CcjNPFn8Y-:-N3f96VR84CcjNPFn8Y0:b-jc76x2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc6a3cfa38e6873331b109c65ff49c27b2a8200/pkg/sql/ttl/ttljob/main_test.go#L18)):*
> ```Go
> 
> 	// Since we are testing multi-tenancy, we need to blank import kvtenantccl.
> 	_ "github.com/cockroachdb/cockroach/pkg/ccl/kvccl/kvtenantccl"
> ```

is this the official way of doing this? or would we want the test file to be in `pkg/ccl`?

maybe this won't matter after @knz 's licensing work?


<!-- Sent from Reviewable.io -->

otan commented:
*[`pkg/sql/ttl/ttljob/main_test.go` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82390#-N3f96VR84CcjNPFn8Y-:-N3fPoCN1F8AkSWtkA0_:btxnb64) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc6a3cfa38e6873331b109c65ff49c27b2a8200/pkg/sql/ttl/ttljob/main_test.go#L18)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

is this the official way of doing this? or would we want the test file to be in `pkg/ccl`?

maybe this won't matter after @knz 's licensing work?
</blockquote></details>

I'm not sure unfortunately :( 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82390)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @knz)

___
*[`pkg/sql/ttl/ttljob/main_test.go` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82390#-N3f96VR84CcjNPFn8Y-:-N3tVgC1CfxOmKkjld4Y:bi4yjpd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc6a3cfa38e6873331b109c65ff49c27b2a8200/pkg/sql/ttl/ttljob/main_test.go#L18)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

I'm not sure unfortunately :( 
</blockquote></details>

The party line is that any tenant testing needs to be moved to a `ccl` package. 


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82390)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[`pkg/sql/ttl/ttljob/main_test.go` line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82390#-N3f96VR84CcjNPFn8Y-:-N3tWmA6ADJWOYW8GBdb:b-o1hjzb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc6a3cfa38e6873331b109c65ff49c27b2a8200/pkg/sql/ttl/ttljob/main_test.go#L18)):*
> maybe this won't matter after @knz 's licensing work?
correct but we're not there yet. so for now this is ok


<!-- Sent from Reviewable.io -->

otan commented:
do we still want this merged & in v22.1?

rafiss commented:
> do we still want this merged & in v22.1?

yes

otan commented:
stamp me :D

rafiss commented on pkg/settings/registry.go:
```diff
@@ -119,6 +119,7 @@ var retiredSettings = map[string]struct{}{
 	"trace.datadog.project":                                            {},
 	"sql.defaults.interleaved_tables.enabled":                          {},
 	"sql.defaults.copy_partitioning_when_deinterleaving_table.enabled": {},
+	"sql.ttl.range_batch_size":                                         {},
```
nit: rebase and put it under the "removed as oof 22.2" comment

otan replied:
Hmm that won't backport nicely, technically this is straight away in 22.1.ill put something in between

rafiss reviewed (approved):
added a nit

otan committed:
ttljob: make TTL work for multi-tenant

Use the appropriate KV abstraction to make ttl work for serverless
clusters.

We can't use `SPLIT AT` with multi-tenant tests, so we are skipping
those for now.

Release note (sql change): The cluster setting
`sql.ttl.range_batch_size` is deprecated.

Release note (sql change): Row-level TTL now works for multi-tenant
clusters.

otan commented:
bors r=rafiss

otan commented:
bors r=rafiss
